packages/api/endpoints/googleToken.js

```diff
@@ -0,0 +1,143 @@
+        auth: oauth2Client
+      }, (err, response) => {
+        if (err) log.error(err);
+        const userData = response.data;
+        // not sure if it's possible to have multiple emails but they are
+        // returned as a list. If users have multiple emails we will have to
+        // scan the users table to see if any match.
+        const userEmail = userData.emails[0].value;
+
+        const u = new User();
+        return u.get({ userName: userEmail })
+          .then(() => {
+            u.update({ userName: userEmail }, { password: accessToken, expires });
+          })
+          .then(() => {
+            if (state) {
+              log.info(`Log info: Redirecting to state: ${state} with token ${accessToken}`);
+              return resp(context, null, 'Redirecting to the specified state', 301, {
+                Location: `${decodeURIComponent(state)}?token=${accessToken}`
+              });
+            }
+            log.info('Log info: No state specified, responding 200');
+            return resp(context, null, JSON.stringify({ token: accessToken }), 200);
+          })
+          .catch((e) => {
+            if (e.message.includes('No record found for')) {
+              return resp(context, new Error('User is not authorized to access this site'));
+            }
+            return resp(context, e);
+          });
+      });
+    });
+  }
+  return resp(context, new Error('Request requires a code'));
+}
+
+/**
+ * `login` is an AWS API Gateway function that redirects to the correct
+ * authentication endpoint with the correct client ID to be used with the API
+ *
+ * @param  {Object} event   - Lambda event object
+ * @param  {Object} context - Lambda context object
+ * @param  {Function} cb    - Lambda callback function
+ * @returns {Function}       Lambda callback function
+ */
+function login(event, context, cb) {
+  // generate a url that asks permissions for Google+ and Google Calendar scopes
+  const scopes = [
+    'https://www.googleapis.com/auth/userinfo.email'
+  ];
+
+  const code = get(event, 'queryStringParameters.code');
+  const state = get(event, 'queryStringParameters.state');
+
+  if (code) {
+    return this.token(event, context);
+  }
+
+  const url = oauth2Client.generateAuthUrl({
+    // 'online' (default) or 'offline' (gets refresh_token)
+    access_type: 'offline',
+
+    // If you only need one scope you can pass it as a string
+    scope: scopes,
+    state: state
+  });
+
+  return cb(null, {
+    statusCode: '301',
+    body: 'Redirecting to Google Login',
+    headers: {
+      Location: url
+    }
+  });
+}
+
+/**
+ * Main handler for the token endpoint.
+ *
+ * @function handler
+ * @param  {Object}   event   - Lambda event payload
+ * @param  {Object}   context - Lambda context - provided by AWS
+ * @param  {Function} cb      - Lambda callback function
+ * @returns {Function}         Calls the `login` or `resp` function.
+ */
+function handler(event, context, cb) {
```

Comment:
Wouldn't it be better if we move the authentication logic for both earthlogin and google to the lib folder and then have a generic token endpoint that decides which one to use based on the configuration?